sycl/doc/CompileTimeProperties.md

```diff
@@ -0,0 +1,680 @@
+# Implementation design for compile time constant properties
+
+This document describes the implementation design for the DPC++ extension
+[SYCL\_EXT\_ONEAPI\_PROPERTY\_LIST][1], which adds a general mechanism for
+specifying properties which are known at compile time.  This extension is not
+itself a feature, but rather a building block that can be incorporated into
+other features.
+
+[1]: <extensions/PropertyList/SYCL_EXT_ONEAPI_property_list.asciidoc>
+
+There are a number of situations where we plan to use compile-time constant
+properties, but this design document does not attempt to address them all.
+Rather, it describes the design for each "category" of use and illustrates each
+category with a specific feature.  For example `accessor` is used to illustrate
+properties that are applied to a kernel argument, but the same technique could
+be used for other variables that are captured as kernel arguments.
+
+In all cases, the goal of this design is to explain how a DPC++ program that
+uses properties is consumed by the device compiler and eventually represented
+in LLVM IR.  This typically involves some logic in the header files which
+results in a C++ annotation that contains the properties.  The device compiler
+front-end is responsible for consuming this annotation and producing some
+corresponding LLVM IR.  One of the goals of this design is to avoid changes to
+the front-end each time we add a new property, so the front-end is not required
+to understand each property it consumes.  Instead, it follows a mechanical
+process for converting properties listed in the C++ annotation into LLVM IR,
+and this mechanical process need not be updated when we add new properties.
+
+Once the information about properties is represented in IR, it is available to
+compiler passes.  For example, the `sycl-post-link` tool might use a property
+in order to perform one of its transformations.  Some properties are consumed
+by the DPC++ compiler, but others are transformed into SPIR-V for use by the
+JIT compiler.  This design document also describes how this SPIR-V
+transformation is done.
+
+
+## Properties on a global variable type
+
+One use for compile-time properties is with types that are used exclusively
+for declaring global variables.  One such example is the
+[SYCL\_EXT\_ONEAPI\_DEVICE\_GLOBAL][2] extension:
+
+[2]: <extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc>
+
+```
+namespace sycl::ext::oneapi {
+
+template <typename T, typename PropertyListT = property_list<>>
+class device_global {/*...*/};
+
+} // namespace sycl::ext::oneapi
+```
+
+The following code illustrates a `device_global` variable that is declared with
+two compile-time properties:
+
+```
+using sycl::ext::oneapi;
+
+device_global<int,
+  property_list_t<
+    device_image_scope::value_t,
+    host_access::value_t<host_access::access::read>>>
+  dm1;
+```
+
+The header file represents these properties with an internal C++ attribute
```

Comment:
Could you please elaborate on why do we need template parameter pack expansion into C++ attribute in headers? As soon as the frontend does the next level of conversion from the attribute which contains list into literal attributes can it process template parameter pack as input?

Reply:
It would be nice to avoid the parameter pack expansion in the C++ attribute, but I don't see how that's possible.  Can you elaborate on your proposal some more?  What does the header file look like in your proposal?

Reply:
My assumption is that the frontend can walk through class template argument list  of the type for the object```(property_list_t<
    device_image_scope::value_t, host_access::value_t<host_access::access::read>>>
  dm1;```). So there is no need to have the intermediate conversion of one of class template arguments into C++ attribute with the list of literals (```[[__sycl_detail__::add_ir_global_variable_attributes()]]```). may be frontend cat generate directly using [GlobalVariable::addAttribute](https://github.com/intel/llvm/pull/4937/files#diff-8b8203c67c38c950b9c19edd0aa137e67e3683c84fc454fe2982ebb19e7a47b1R129), [Function::addFnAttr](https://github.com/intel/llvm/pull/4937/files#diff-8b8203c67c38c950b9c19edd0aa137e67e3683c84fc454fe2982ebb19e7a47b1R338), [Function::addParamAttrs](https://github.com/intel/llvm/pull/4937/files#diff-8b8203c67c38c950b9c19edd0aa137e67e3683c84fc454fe2982ebb19e7a47b1R238) directly from class template arguments.

Reply:
I think this would require us to change the front-end whenever the headers use property_list in a new class.  For example, if we add `annotated_ptr` (which takes a property list), the front-end would need to be changed to recognize the `annotated_ptr` type (by name), and then it would need to identify which template parameter corresponded to the property list and then parse the properties from that template parameter.  One of the goals of this design is to avoid the need to change the front-end each time we need a new property (or add a new class that uses properties).

With the proposed design, the front-end needs to recognize the C++ attributes, but it does not need to recognize the C++ class that uses these attributes.  We can add new classes that use properties without any additional change to the front-end.

Reply:
The other option is to detect the specific type of class template argument (property_list_t). if it is property_list_t transform its content into C++ attributes for the specific object.

Reply:
There are two downsides to this that I see:

1. The front-end would have a dependency on the internal class structure of each property in the property list.  This might mean that we would require front-end changes if we added a new property whose implementation details was even slightly different from an existing property.  This seems less flexible to me that using a C++ attribute with a parameter pack.  With the C++ attribute approach, the front-end doesn't care at all about the structure of a property.  The only dependency between the headers and the front-end is the definition of the C++ attribute.

2. The headers won't have any way to indicate which type of IR representation to use.  For example, the header might want to represent some `accessor` properties as IR kernel parameter attribute and represent other properties as IR annotations.

Reply:
My few cents for this discussion. I find new specific attributes approach to be much clearer, this-repository-developer-friendly (just consider [our implementation of buffer_location accessor property](https://github.com/intel/llvm/pull/2166) and then imagine, that we have tons of possible properties for each special class just for a singe device) and external-developer-friendly (processing special types can make our FE incompatible with other library implementations). Also each time processing a special SYCL type might be compile-time costly.

Reply:
As a side question, why can't we just use `clang::annotation` attribute, which is now (for about a year) a parameter pack attribute as well? If we want to distinguish headers-added attributes from user-added, we can add some magic string to as the first parameter, so this example code:
```
template <typename T, typename ...Props>
class annotated_ptr<T, property_list<Props...>> {
  T *ptr
#ifdef __SYCL_DEVICE_ONLY__
  [[__sycl_detail__::add_ir_member_annotation(
    "foo",    // Name of first property
    "bar",    // Name of second property
    "",       // First property has no parameter
    32        // Value of second property
    )]]
#endif
  ;
 public:
  annotated_ptr(T *p) : ptr(p) {}
};
```

would be transformed to:
```
template <typename T, typename ...Props>
class annotated_ptr<T, property_list<Props...>> {
  T *ptr
#ifdef __SYCL_DEVICE_ONLY__
  [[clang::annotation(
    "__sycl_detail_ptr_annotation" // Magic string
    "foo",    // Name of first property
    "bar",    // Name of second property
    "",       // First property has no parameter
    32        // Value of second property
    )]]
#endif
  ;
 public:
  annotated_ptr(T *p) : ptr(p) {}
};
```

?

Reply:
@MrSidims - Just to chime in, I cannot find any `clang::annotation` so I assume you're referring to `clang::annotate`? It does not seem to handle parameter packs currently (see https://godbolt.org/z/z8WsKb8Ys), which I believe is a more general limitation that also affects the new attributes introduced here.

Reply:
Hm, godbolt result surprises me. https://reviews.llvm.org/D88645 made me thinking, that the example should be compilable. Unless the patch got reverted. May be @AaronBallman can point out what I'm missing.

Reply:
And it does look like that it has at minimum one bug, compilation of an easier example results in a compiler crash: https://godbolt.org/z/c4Wd87nsK

Reply:
> Yeah, I was referring to `clang::annotate`, thanks :) . Hm, godbolt result surprises me. https://reviews.llvm.org/D88645 made me thinking, that the example should be compilable. Unless the patch got reverted. May be @AaronBallman can point out what I'm missing.

I think the mentioned patch just made the arguments variadic, but the problem seems to be that the parser isn't ready for the parameter pack ellipsis in attribute arguments. I don't know exactly what C++ says about this, but playing around with it a bit I believe it is relatively easy to amend.

> And it does look like that it has at minimum one bug, compilation of an easier example results in a compiler crash: https://godbolt.org/z/c4Wd87nsK

I am not sure that should compile, but at the very least it should result in a readable error not a segfault. Nice find! 😄

Reply:
The following appears to make the first example work: https://github.com/steffenlarsen/llvm/commit/ac8f157c3b68d3c3f563aa8f6dbdf166f297175c

I may have missed some caveats and more testing would be needed, but it is a start.

Reply:
Thanks for pointing this out @MrSidims.  I knew about `[[clang::annotate()]]`, but I didn't realize it could take a variadic parameter list.  Apparently, parameters after the first are emitted as the 5th parameter to `@llvm.ptr.annotation`, which is not documented in the [LLVM Language Reference Manual](https://llvm.org/docs/LangRef.html#llvm-ptr-annotation-intrinsic).  Therefore, something like this:

```
template <typename T, typename ...Props>
class annotated_ptr<T, property_list<Props...>> {
  T *ptr
#ifdef __SYCL_DEVICE_ONLY__
  [[clang::annotation(
    "sycl-properties" // Magic string
    "foo",    // Name of first property
    "bar",    // Name of second property
    nullptr,  // First property has no parameter
    32        // Value of second property
    )]]
#endif
  ;
 public:
  annotated_ptr(T *p) : ptr(p) {}
};
```

Becomes:

```
@.str = private unnamed_addr constant [16 x i8] c"sycl-properties\00", section "llvm.metadata"
@.str.1 = private unnamed_addr constant [9 x i8] c"file.cpp\00", section "llvm.metadata"
@.str.2 = private unnamed_addr constant [4 x i8] c"foo\00", align 1
@.str.3 = private unnamed_addr constant [4 x i8] c"bar\00", align 1
@.args = private unnamed_addr constant { [4 x i8]*, [4 x i8]*, i8*, i32 } { [4 x i8]* @.str.2, [4 x i8]* @.str.3, i8* null, i32 32 }, section "llvm.metadata"

; ...

%1 = call i8* @llvm.ptr.annotation.p0i8(i8* nonnull %0,
  i8* getelementptr inbounds ([16 x i8], [16 x i8]* @.str, i64 0, i64 0), 
  i8* getelementptr inbounds ([9 x i8], [9 x i8]* @.str.1, i64 0, i64 0), 
  i32 3, 
  i8* bitcast ({ [4 x i8]*, [4 x i8]*, i8*, i32 }* @.args to i8*)) 
```

That actually seems nicer than what I originally proposed because the SYCL properties aren't all combined into a single string.  Unless this undocumented 5th parameter to `@llvm.ptr.annotation` has some existing semantic that we are abusing, then I agree that this is a better strategy than inventing a new C++ attribute `[[__sycl_detail__::add_ir_member_annotation()]]`.

Reply:
There was an attempt to update the LangRef https://reviews.llvm.org/D96646 , but it doesn't cover every case of annotation usage. I've opened some time ago an internal tracker to fix LangRef and planned to do it myself, but I haven't prioritized it yet, leaving it on the second to last position in the tasks queue.

Reply:
I decided to keep the proposed C++ attribute `[[__sycl_detail__::add_ir_member_annotation()]]`, but I changed the way it represents the properties in IR to be the same as what `[[clang::annotate()]]` does.  There are two reasons I want to add  `[[__sycl_detail__::add_ir_member_annotation()]]` rather than use `[[clang::annotate()]]`:

* The main reason is because I want to introduce a new optional parameter that allows filtering of the SYCL properties.  See the section "Filtering properties" for an explanation of this parameter.

* A secondary reason, is that I'd like the front-end to reorder the parameters when constructing the tuple that is passed as the fifth metadata argument to `@llvm.ptr.annotation`.  The behavior of `[[clang::annotate()]]` is to generate the elements of the tuple in the same order as they appear in the parameter list to `[[clang::annotate()]]`.  However, this would result in a tuple like `{"prop1-name", "prop2-name", prop1-value, prop2-value}`.  Instead, I think the IR would be easier to consume if the order was `{"prop1-name", prop1-value, "prop2-name", prop2-value}`.

I made this change in 44763693264bc91488fc5c8772b61f0cc5d9210b.

Reply:
> Thanks for pointing this out @MrSidims. I knew about `[[clang::annotate()]]`, but I didn't realize it could take a variadic parameter list. Apparently, parameters after the first are emitted as the 5th parameter to `@llvm.ptr.annotation`, which is not documented in the [LLVM Language Reference Manual](https://llvm.org/docs/LangRef.html#llvm-ptr-annotation-intrinsic). 

Perhaps you should follow the triSYCL project? :-) @Ralender upstreamed this variadic extension to Clang/LLVM because I disliked so much a spam of attributes. ;-)